### PR TITLE
[PowerPC] Combine sub within setcc back to sext

### PR DESCRIPTION
**Pull request by ecnelises**
InstCombine does below transformation:

```llvm
; bool cmp(i64 a) { return a != (i64)(iN)a; }
%0 = add i64 %a, -2^(N-1)
%1 = icmp ult i64 %0, -2^N
ret i1 %1

; bool cmp(i64 a) { return a == (i64)(iN)a; }
%0 = add i64 %a, 2^(N-1)
%1 = icmp ult i64 %0, 2^N
ret i1 %1
```

On PowerPC, it's profitable to combine this add-cmpult back to original trunc-ext-cmpeq.

### CODE REVIEW

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp [hunk 1]
```diff
@@ -14428,15 +14431,53 @@ SDValue PPCTargetLowering::combineSetCC(SDNode *N,
...
+          SDValue Conv = DAG.getSExtOrTrunc(
+              DAG.getSExtOrTrunc(LHS.getOperand(0), DL, MVT::i8), DL, OpVT);
+          return DAG.getSetCC(DL, VT, LHS.getOperand(0), Conv, ISD::SETNE);
+        }
+      }
+    } else if (LHS.getOpcode() == ISD::SRL &&
+               LHS.getOperand(0).getOpcode() == ISD::ADD &&
+               isa<ConstantSDNode>(LHS.getOperand(1)) &&
+               isa<ConstantSDNode>(LHS.getOperand(0).getOperand(1))) {
+      if (RHSVal == 65535 &&
```
Comment (bzEq) [inline llvm/lib/Target/PowerPC/PPCISelLowering.cpp:14468]:
```suggestion
      if (RHSVal == 0xff &&
```

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp [hunk 2]
```diff
@@ -14428,15 +14431,53 @@ SDValue PPCTargetLowering::combineSetCC(SDNode *N,
...
       return DAG.getSetCC(DL, VT, Add, DAG.getConstant(0, DL, OpVT), CC);
     }
   }
 
+  if (CC == ISD::SETULT && isa<ConstantSDNode>(RHS)) {
+    uint64_t RHSVal = cast<ConstantSDNode>(RHS)->getZExtValue();
+    if (LHS.getOpcode() == ISD::ADD && isa<ConstantSDNode>(LHS.getOperand(1))) {
+      uint64_t Addend = cast<ConstantSDNode>(LHS.getOperand(1))->getZExtValue();
+      if (OpVT == MVT::i64) {
+        // (a-2^(M-1)) => sext(trunc(a, M), 64)
```
Comment (bzEq) [inline llvm/lib/Target/PowerPC/PPCISelLowering.cpp:14444]: When is this equation hold?

**Comment by bzEq** [comment]
Please provide more description in PR summary.

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp [hunk 3]
```diff
@@ -14428,15 +14431,52 @@ SDValue PPCTargetLowering::combineSetCC(SDNode *N,
...
   }
 
+  if (CC == ISD::SETULT && isa<ConstantSDNode>(RHS)) {
+    uint64_t RHSVal = cast<ConstantSDNode>(RHS)->getZExtValue();
+    if (LHS.getOpcode() == ISD::ADD && isa<ConstantSDNode>(LHS.getOperand(1))) {
+      uint64_t Addend = cast<ConstantSDNode>(LHS.getOperand(1))->getZExtValue();
+      if (OpVT == MVT::i64) {
+        uint64_t ShiftVal = ~Addend + 1;
+        uint64_t CmpVal = ~RHSVal + 1;
+        if (isPowerOf2_64(ShiftVal) && ShiftVal << 1 == CmpVal) {
```
Comment (bzEq) [inline llvm/lib/Target/PowerPC/PPCISelLowering.cpp:14446]: Add comment for the DAG pattern found. Better provide alive2 prove in the PR summary.

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp [hunk 4]
```diff
@@ -14428,15 +14431,52 @@ SDValue PPCTargetLowering::combineSetCC(SDNode *N,
...
       return DAG.getSetCC(DL, VT, Add, DAG.getConstant(0, DL, OpVT), CC);
     }
   }
 
+  if (CC == ISD::SETULT && isa<ConstantSDNode>(RHS)) {
+    uint64_t RHSVal = cast<ConstantSDNode>(RHS)->getZExtValue();
+    if (LHS.getOpcode() == ISD::ADD && isa<ConstantSDNode>(LHS.getOperand(1))) {
+      uint64_t Addend = cast<ConstantSDNode>(LHS.getOperand(1))->getZExtValue();
+      if (OpVT == MVT::i64) {
+        uint64_t ShiftVal = ~Addend + 1;
```
Comment (bzEq) [inline llvm/lib/Target/PowerPC/PPCISelLowering.cpp:14444]:
```suggestion
        uint64_t ShiftVal = -Addend;
```

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp [hunk 5]
```diff
@@ -14428,15 +14431,52 @@ SDValue PPCTargetLowering::combineSetCC(SDNode *N,
...
+        uint64_t CmpVal = ~RHSVal + 1;
```
Comment (bzEq) [inline llvm/lib/Target/PowerPC/PPCISelLowering.cpp:14445]:
```suggestion
        uint64_t CmpVal = -RHSVal;
```

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp (hunk 4 above)
Comment (ecnelises) [reply llvm/lib/Target/PowerPC/PPCISelLowering.cpp:14444]: negating an unsigned variable is UB?

**Comment by ecnelises** [comment]
The patterns are

```llvm
; ult (add x -0x80000000) -0x100000000 -> ne x (sext:i64 (trunc:i32 x))
; ult (add x -0x8000) -0x10000 -> ne x (sext:i64 (trunc:i16 x))
; ult (add x -0x80) -0x100 -> ne x (sext:i64 (trunc:i8 x))
; ult (add x -0x80) -0x100 -> ne x (sext:i32 (trunc:i16 x))
; ult (add x -0x80) -0x100 -> ne x (sext:i16 (trunc:i8 x))
; ult (srl (add x -0x8000) 16) 0xffff -> ne x (sext:i32 (trunc:i16 x))
```

Only the last one differs from the rest, because it's been transformed in generic DAG combiner. (https://alive2.llvm.org/ce/z/jAHgKM ) The rest are just reverting transformation of InstCombine.

File: llvm/test/CodeGen/PowerPC/setcc-to-sub.ll [hunk 6]
```diff
@@ -171,12 +164,13 @@ entry:
...
+; CHECK-NEXT:    extsb 3, 3
 ; CHECK-NEXT:    clrlwi 3, 3, 16
-; CHECK-NEXT:    ori 4, 4, 256
-; CHECK-NEXT:    add 3, 3, 4
-; CHECK-NEXT:    rldicl 3, 3, 1, 63
+; CHECK-NEXT:    xor 3, 4, 3
+; CHECK-NEXT:    cntlzw 3, 3
+; CHECK-NEXT:    srwi 3, 3, 5
+; CHECK-NEXT:    xori 3, 3, 1
 ; CHECK-NEXT:    blr
```
Comment (ecnelises) [inline llvm/test/CodeGen/PowerPC/setcc-to-sub.ll:174]: Will investigate about the regression.

Comment (ecnelises) [reply llvm/test/CodeGen/PowerPC/setcc-to-sub.ll:172]: This should not be regression. Parameter should have `signext`.

**Pull request closed without merging**